[FLINK-15391][hive] DATE and TIMESTAMP partition columns don't work

lirui-apache opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Travis CI to do that following [this guide](https://flink.apache.org/contributing/contribute-code.html#open-a-pull-request).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

Fix issues of read/write DATE and TIMESTAMP partition columns.


## Brief change log

  - When reading DATE/TIMESTAMP partition columns, make sure the value objects are converted to corresponding Flink objects.
  - When writing DATE/TIMESTAMP partition columns, make sure we convert to Hive objects before extracting the string values.
  - Add new tests


## Verifying this change

New test cases

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? NA

lirui-apache commented:
cc @JingsongLi @zjuwangg @bowenli86 @xuefuz

bowenli86 commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HivePartitionComputer.java:
```diff
@@ -0,0 +1,80 @@
+			if (field != null) {
+				field = partColConversions[i].toHiveObject(field);
+			}
+			String partitionValue = field != null ? field.toString() : null;
```
minor: these can be consolidated on just `field != null`?

bowenli86 commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HivePartitionComputer.java:
```diff
@@ -0,0 +1,80 @@
+				field = partColConversions[i].toHiveObject(field);
+			}
+			String partitionValue = field != null ? field.toString() : null;
+			if (partitionValue == null || "".equals(partitionValue)) {
```
minor: StringUtils.isNullOrEmptyString(partitionValue)?

lirui-apache replied:
I don't think we have the `StringUtils.isNullOrEmptyString` util. We do have `StringUtils.isNullOrWhitespaceOnly` but it's not exactly the same as the logic here.

JingsongLi replied:
`org.apache.commons.lang3.StringUtils.isEmpty` is OK.

JingsongLi commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HivePartitionComputer.java:
```diff
@@ -0,0 +1,80 @@
+
+	private final HiveShim hiveShim;
+	private final DataType[] columnTypes;
+	private HiveObjectConversion[] partColConversions;
```
Can we make `HiveObjectConversion` serializable?

JingsongLi commented on flink-connectors/flink-connector-hive/src/main/java/org/apache/flink/connectors/hive/HiveTableSource.java:
```diff
@@ -286,7 +293,15 @@ private Object restorePartitionValueFromFromType(String valStr, DataType type) {
 				return Double.valueOf(valStr);
 			case DATE:
-				return Date.valueOf(valStr);
+				return HiveInspectors.toFlinkObject(
```
You can  fix `AbstractOrcColumnVector.createHiveVectorFromConstant` too.

lirui-apache commented:
@bowenli86 @JingsongLi Thanks for the review. I rebased and addressed your comments.
I updated `ExpressionExtractor` to convert to Hive object when visiting a literal value. But since currently we don't support prune DATE/TIMESTAMP partition columns, we can't add test to verify that change.

bowenli86 commented:
LGTM, merging.

Regarding "we don't support prune DATE/TIMESTAMP partition columns, we can't add test to verify that change.", is that a bug or did we miss it in the first place? I think it should be part of the story of fully supporting partition pruning in HiveCatalog. Are we able to do it in 1.10?
